File: java/src/main/java/com/sasteval/vuln/medium/LdapInjection.java
package com.sasteval.vuln.medium;

import javax.naming.Context;
import javax.naming.NamingEnumeration;
import javax.naming.directory.DirContext;
import javax.naming.directory.InitialDirContext;
import javax.naming.directory.SearchControls;
import javax.naming.directory.SearchResult;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Hashtable;

/**
 * CWE-90: LDAP Injection (Cross-Method)
 * VULNERABILITY: User input is concatenated into LDAP search filter.
 */
public class LdapInjection extends HttpServlet {

    private NamingEnumeration<SearchResult> searchLdap(String username) throws Exception {
        Hashtable<String, String> env = new Hashtable<>();
        env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");
        env.put(Context.PROVIDER_URL, "ldap://localhost:389");

        DirContext ctx = new InitialDirContext(env);
        SearchControls controls = new SearchControls();
        controls.setSearchScope(SearchControls.SUBTREE_SCOPE);

        // VULN: User input concatenated into LDAP filter without escaping
        String filter = "(uid=" + username + ")";
        return ctx.search("dc=example,dc=com", filter, controls);
    }

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {

        String username = request.getParameter("username");
        response.setContentType("application/json");

        try {
            NamingEnumeration<SearchResult> results = searchLdap(username);
            StringBuilder sb = new StringBuilder("[");
            while (results.hasMore()) {
                SearchResult result = results.next();
                sb.append("\"").append(result.getNameInNamespace()).append("\",");
            }
            if (sb.length() > 1) sb.setLength(sb.length() - 1);
            sb.append("]");
            response.getWriter().println(sb.toString());

        } catch (Exception e) {
            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
            response.getWriter().println("{\"error\": \"LDAP query failed\"}");
        }
    }
}

File: dotnet/Vuln/Medium/LdapInjection.cs
using System.DirectoryServices;

namespace SastEval.Vuln.Medium;

/// <summary>
/// CWE-90: LDAP Injection - User input concatenated into LDAP search filter.
/// </summary>
public static class LdapInjection
{
    private static string SearchLdap(string username)
    {
        // VULNERABLE: User input directly concatenated into LDAP filter
        var filter = "(uid=" + username + ")";

        try
        {
            var entry = new DirectoryEntry("LDAP://localhost:389");
            var searcher = new DirectorySearcher(entry)
            {
                Filter = filter
            };

            var result = searcher.FindOne();
            return result?.GetDirectoryEntry().Name ?? "User not found";
        }
        catch (Exception ex)
        {
            return $"LDAP search error: {ex.Message}";
        }
    }

    public static IResult Handle(HttpContext context)
    {
        var username = context.Request.Query["username"].ToString();

        if (string.IsNullOrEmpty(username))
        {
            return Results.BadRequest("Missing 'username' query parameter");
        }

        var result = SearchLdap(username);
        return Results.Text(result);
    }
}
